File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/activity/record/MainActivity.java
package com.blogspot.e_kanivets.moneytracker.activity.record;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.ListView;
import android.widget.TextView;

import com.blogspot.e_kanivets.moneytracker.controller.CurrencyController;
import com.blogspot.e_kanivets.moneytracker.R;
import com.blogspot.e_kanivets.moneytracker.activity.ReportActivity;
import com.blogspot.e_kanivets.moneytracker.activity.base.BaseDrawerActivity;
import com.blogspot.e_kanivets.moneytracker.adapter.RecordAdapter;
import com.blogspot.e_kanivets.moneytracker.controller.FormatController;
import com.blogspot.e_kanivets.moneytracker.controller.PeriodController;
import com.blogspot.e_kanivets.moneytracker.controller.data.AccountController;
import com.blogspot.e_kanivets.moneytracker.controller.data.ExchangeRateController;
import com.blogspot.e_kanivets.moneytracker.controller.PreferenceController;
import com.blogspot.e_kanivets.moneytracker.controller.data.RecordController;
import com.blogspot.e_kanivets.moneytracker.entity.data.Account;
import com.blogspot.e_kanivets.moneytracker.entity.data.Record;
import com.blogspot.e_kanivets.moneytracker.entity.Period;
import com.blogspot.e_kanivets.moneytracker.report.ReportMaker;
import com.blogspot.e_kanivets.moneytracker.report.record.IRecordReport;
import com.blogspot.e_kanivets.moneytracker.ui.AppRateDialog;
import com.blogspot.e_kanivets.moneytracker.ui.PeriodSpinner;
import com.blogspot.e_kanivets.moneytracker.ui.presenter.ShortSummaryPresenter;
import com.blogspot.e_kanivets.moneytracker.util.AnswersProxy;

import java.util.Collections;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.OnClick;
import butterknife.OnItemClick;

public class MainActivity extends BaseDrawerActivity {
    @SuppressWarnings("unused")
    private static final String TAG = "MainActivity";

    private static final int REQUEST_ACTION_RECORD = 6;

    private List<Record> recordList;
    private Period period;

    @Inject
    RecordController recordController;
    @Inject
    ExchangeRateController rateController;
    @Inject
    AccountController accountController;
    @Inject
    CurrencyController currencyController;
    @Inject
    PreferenceController preferenceController;
    @Inject
    PeriodController periodController;
    @Inject
    FormatController formatController;

    private ShortSummaryPresenter summaryPresenter;

    @BindView(R.id.list_view)
    ListView listView;
    @BindView(R.id.spinner_period)
    PeriodSpinner spinner;

    TextView tvDefaultAccountTitle;
    TextView tvDefaultAccountSum;
    TextView tvCurrency;

    @Override
    protected int getContentViewId() {
        return R.layout.activity_main;
    }

    @Override
    protected boolean initData() {
        super.initData();
        getAppComponent().inject(MainActivity.this);

        preferenceController.addLaunchCount();
        summaryPresenter = new ShortSummaryPresenter(MainActivity.this);

        return super.initData();
    }

    @Override
    protected void initViews() {
        super.initViews();

        setTitle(R.string.title_records);

        tvDefaultAccountTitle = (TextView) navigationView.getHeaderView(0).findViewById(R.id.tv_default_account_title);
        tvDefaultAccountSum = (TextView) navigationView.getHeaderView(0).findViewById(R.id.tv_default_account_sum);
        tvCurrency = (TextView) navigationView.getHeaderView(0).findViewById(R.id.tv_currency);

        if (preferenceController.checkRateDialog()) showAppRateDialog();

        registerForContextMenu(listView);

        View summaryView = summaryPresenter.create(true);
        listView.addHeaderView(summaryView);
        summaryView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                showReport();
            }
        });

        spinner.setPeriodSelectedListener(new PeriodSpinner.OnPeriodSelectedListener() {
            @Override
            public void onPeriodSelected(Period period) {
                MainActivity.this.period = period;
                periodController.writeLastUsedPeriod(period);
                update();
            }
        });
        spinner.setPeriod(periodController.readLastUsedPeriod());
    }

    @OnItemClick(R.id.list_view)
    public void editRecord(int position) {
        AnswersProxy.get().logButton("Edit Record");
        // Minus one because of list view's header view
        Record record = recordList.get(position - 1);
        startAddRecordActivity(record, AddRecordActivity.Mode.MODE_EDIT, record.getType());
    }

    @OnClick(R.id.btn_add_expense)
    public void addExpense() {
        AnswersProxy.get().logButton("Add Expense");
        startAddRecordActivity(null, AddRecordActivity.Mode.MODE_ADD, Record.TYPE_EXPENSE);
    }

    @OnClick(R.id.btn_add_income)
    public void addIncome() {
        AnswersProxy.get().logButton("Add Income");
        startAddRecordActivity(null, AddRecordActivity.Mode.MODE_ADD, Record.TYPE_INCOME);
    }

    public void showReport() {
        AnswersProxy.get().logButton("Show Report");
        Intent intent = new Intent(MainActivity.this, ReportActivity.class);
        intent.putExtra(ReportActivity.KEY_PERIOD, period);
        startActivity(intent);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        if (resultCode == AppCompatActivity.RESULT_OK) {
            switch (requestCode) {
                case REQUEST_ACTION_RECORD:
                    update();
                    break;

                case REQUEST_BACKUP:
                    getAppComponent().inject(MainActivity.this);
                    update();
                    break;

                default:
                    break;
            }
        }
    }

    @Override
    protected void update() {
        recordList = recordController.getRecordsForPeriod(period);
        Collections.reverse(recordList);

        listView.setAdapter(new RecordAdapter(MainActivity.this, recordList));

        String currency = currencyController.readDefaultCurrency();

        ReportMaker reportMaker = new ReportMaker(rateController);
        IRecordReport report = reportMaker.getRecordReport(currency, period, recordList);
        summaryPresenter.update(report, currency, reportMaker.currencyNeeded(currency, recordList));

        fillDefaultAccount();
    }

    private void showAppRateDialog() {
        AnswersProxy.get().logEvent("Show App Rate Dialog");
        AppRateDialog dialog = new AppRateDialog(MainActivity.this);
        dialog.setCanceledOnTouchOutside(false);
        dialog.show();
    }

    private void startAddRecordActivity(Record record, AddRecordActivity.Mode mode, int type) {
        Intent intent = new Intent(MainActivity.this, AddRecordActivity.class);
        intent.putExtra(AddRecordActivity.KEY_RECORD, record);
        intent.putExtra(AddRecordActivity.KEY_MODE, mode);
        intent.putExtra(AddRecordActivity.KEY_TYPE, type);
        startActivityForResult(intent, REQUEST_ACTION_RECORD);
    }

    @SuppressLint("SetTextI18n")
    private void fillDefaultAccount() {
        Account defaultAccount = accountController.readDefaultAccount();
        if (defaultAccount == null) return;

        tvDefaultAccountTitle.setText(defaultAccount.getTitle());
        tvDefaultAccountSum.setText(formatController.formatAmount(defaultAccount.getFullSum()));
        tvCurrency.setText(defaultAccount.getCurrency());
    }
}

File: app/src/main/java/com/blogspot/e_kanivets/moneytracker/activity/record/MainActivity.kt
package com.blogspot.e_kanivets.moneytracker.activity.record

import android.content.Intent
import android.support.v7.app.AppCompatActivity
import android.widget.TextView
import com.blogspot.e_kanivets.moneytracker.R
import com.blogspot.e_kanivets.moneytracker.activity.ReportActivity
import com.blogspot.e_kanivets.moneytracker.activity.base.BaseDrawerActivity
import com.blogspot.e_kanivets.moneytracker.adapter.RecordAdapter
import com.blogspot.e_kanivets.moneytracker.controller.CurrencyController
import com.blogspot.e_kanivets.moneytracker.controller.FormatController
import com.blogspot.e_kanivets.moneytracker.controller.PeriodController
import com.blogspot.e_kanivets.moneytracker.controller.PreferenceController
import com.blogspot.e_kanivets.moneytracker.controller.data.AccountController
import com.blogspot.e_kanivets.moneytracker.controller.data.ExchangeRateController
import com.blogspot.e_kanivets.moneytracker.controller.data.RecordController
import com.blogspot.e_kanivets.moneytracker.entity.Period
import com.blogspot.e_kanivets.moneytracker.entity.RecordItem
import com.blogspot.e_kanivets.moneytracker.entity.data.Record
import com.blogspot.e_kanivets.moneytracker.report.ReportMaker
import com.blogspot.e_kanivets.moneytracker.ui.AppRateDialog
import com.blogspot.e_kanivets.moneytracker.util.AnswersProxy
import com.blogspot.e_kanivets.moneytracker.util.RecordItemsBuilder
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.content_main.*
import javax.inject.Inject

class MainActivity : BaseDrawerActivity() {

    private lateinit var recordList: List<Record>
    private var recordItems: List<RecordItem> = listOf()
    private lateinit var period: Period
    private lateinit var recordAdapter: RecordAdapter

    @Inject
    lateinit var recordController: RecordController
    @Inject
    lateinit var rateController: ExchangeRateController
    @Inject
    lateinit var accountController: AccountController
    @Inject
    lateinit var currencyController: CurrencyController
    @Inject
    lateinit var preferenceController: PreferenceController
    @Inject
    lateinit var periodController: PeriodController
    @Inject
    lateinit var formatController: FormatController

    private lateinit var tvDefaultAccountTitle: TextView
    private lateinit var tvDefaultAccountSum: TextView
    private lateinit var tvCurrency: TextView

    override fun getContentViewId(): Int = R.layout.activity_main

    override fun initData(): Boolean {
        super.initData()
        appComponent.inject(this)

        preferenceController.addLaunchCount()

        return super.initData()
    }

    override fun initViews() {
        super.initViews()

        setTitle(R.string.title_records)

        if (preferenceController.checkRateDialog()) showAppRateDialog()

        tvDefaultAccountTitle = navigationView.getHeaderView(0).findViewById(R.id.tvDefaultAccountTitle)
        tvDefaultAccountSum = navigationView.getHeaderView(0).findViewById(R.id.tvDefaultAccountSum)
        tvCurrency = navigationView.getHeaderView(0).findViewById(R.id.tvCurrency)

        recordAdapter = RecordAdapter(this, listOf(), true) { position ->
            if (position == 0) showReport()
            else editRecord(position)
        }
        recyclerView.adapter = recordAdapter

        spinner.setPeriodSelectedListener { period ->
            this.period = period
            periodController.writeLastUsedPeriod(period)
            update()
        }

        spinner.setPeriod(periodController.readLastUsedPeriod())

        btnAddExpense.setOnClickListener { addExpense() }
        btnAddIncome.setOnClickListener { addIncome() }
    }

    private fun editRecord(position: Int) {
        AnswersProxy.get().logButton("Edit Record")

        val record = recordList[position - 1 - getCountHeadersItems(position - 1)]
        startAddRecordActivity(record, AddRecordActivity.Mode.MODE_EDIT, record.type)
    }

    private fun addExpense() {
        AnswersProxy.get().logButton("Add Expense")
        startAddRecordActivity(null, AddRecordActivity.Mode.MODE_ADD, Record.TYPE_EXPENSE)
    }

    private fun addIncome() {
        AnswersProxy.get().logButton("Add Income")
        startAddRecordActivity(null, AddRecordActivity.Mode.MODE_ADD, Record.TYPE_INCOME)
    }

    private fun showReport() {
        AnswersProxy.get().logButton("Show Report")
        val intent = Intent(this, ReportActivity::class.java)
        intent.putExtra(ReportActivity.KEY_PERIOD, period)
        startActivity(intent)
    }

    public override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)

        if (resultCode == AppCompatActivity.RESULT_OK) {
            when (requestCode) {
                REQUEST_ACTION_RECORD -> update()

                REQUEST_BACKUP -> {
                    appComponent.inject(this)
                    update()
                }

                else -> {
                }
            }
        }
    }

    override fun update() {
        recordList = recordController.getRecordsForPeriod(period)
        recordList = recordList.reversed()
        recordItems = RecordItemsBuilder().getRecordItems(recordList)

        val currency = currencyController.readDefaultCurrency()

        val reportMaker = ReportMaker(rateController)
        val report = reportMaker.getRecordReport(currency, period, recordList)

        recordAdapter.setRecords(recordItems, report, currency, reportMaker.currencyNeeded(currency, recordList))

        fillDefaultAccount()
    }

    private fun getCountHeadersItems(position: Int): Int {
        var countHeadersItems = 0
        for (inOfData in 0 until position) {
            if (recordItems[inOfData] is RecordItem.Header) {
                countHeadersItems++
            }
        }
        return countHeadersItems
    }

    private fun showAppRateDialog() {
        AnswersProxy.get().logEvent("Show App Rate Dialog")
        val dialog = AppRateDialog(this)
        dialog.setCanceledOnTouchOutside(false)
        dialog.show()
    }

    private fun startAddRecordActivity(record: Record?, mode: AddRecordActivity.Mode, type: Int) {
        val intent = Intent(this, AddRecordActivity::class.java)
        intent.putExtra(AddRecordActivity.KEY_RECORD, record)
        intent.putExtra(AddRecordActivity.KEY_MODE, mode)
        intent.putExtra(AddRecordActivity.KEY_TYPE, type)
        startActivityForResult(intent, REQUEST_ACTION_RECORD)
    }

    private fun fillDefaultAccount() {
        val defaultAccount = accountController.readDefaultAccount() ?: return

        tvDefaultAccountSum.text = defaultAccount.title
        tvDefaultAccountSum.text = formatController.formatAmount(defaultAccount.fullSum)
        tvCurrency.text = defaultAccount.currency
    }

    companion object {
        private const val REQUEST_ACTION_RECORD = 6
    }

}